Allow additional configuration metadata to VM during deployment

dhlaluku opened the pull request:
## Description
<!--- Describe your changes in detail -->
Currently, CloudStack does not allow additional configuration metadata to guest VMs during deployment. This feature aims to provide operators the functionality to pass arbitrary additional configuration metadata that will be included everytime the guest VM is started.

<!-- For new features, provide link to FS, dev ML discussion etc. -->
<!-- In case of bug fix, the expected and actual behaviours, steps to reproduce. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## GitHub Issue/PRs
<!-- If this PR is to fix an issue or another PR on GH, uncomment the section and provide the id of issue/PR -->
<!-- When "Fixes: #<id>" is specified, the issue/PR will automatically be closed when this PR gets merged -->
<!-- For addressing multiple issues/PRs, use multiple "Fixes: #<id>" -->

<!-- Fixes: # -->

## Screenshots (if appropriate):

## How Has This Been Tested?

<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [CONTRIBUTING](https://github.com/apache/cloudstack/blob/master/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
Testing
- [ ] I have added tests to cover my changes.
- [ ] All relevant new and existing integration tests have passed.
- [ ] A full integration testsuite with all test that can run on my environment has passed.

yadvr commented on api/src/main/java/org/apache/cloudstack/api/command/user/vm/DeployVMCmd.java:
```diff
@@ -482,6 +485,10 @@ public String getKeyboard() {
         return dataDiskTemplateToDiskOfferingMap;
     }
 
+    public String getExtraConfig() {
```
The extraconfig is not consumed by the API.

dhlaluku replied:
[WIP] It is consumed by the Service Layer (UserVmManegerImpl) for passing extra config, will add nullity validations/checks.

yadvr commented on api/src/main/java/org/apache/cloudstack/api/command/user/vm/UpdateVMCmd.java:
```diff
@@ -221,6 +224,10 @@ public boolean isCleanupDetails(){
 
+    public String getExtraConfig() {
+        return extraConfig;
+    }
```
Extra config not used.

dhlaluku replied:
Consumed by Service Layer

yadvr commented on api/src/main/java/org/apache/cloudstack/api/ApiConstants.java:
```diff
@@ -725,6 +725,8 @@
     public static final String EXITCODE = "exitcode";
     public static final String TARGET_ID = "targetid";
 
+    public static final String EXTRA_CONFIG = "extraconfig";
```
Move this alphabetically up in the list.

dhlaluku replied:
done

yadvr commented on api/src/main/java/com/cloud/agent/api/to/VirtualMachineTO.java:
```diff
@@ -74,6 +74,16 @@
+        this.extraConfig = extraConfig;
+    }
+
+    String extraConfig = null;
```
Move this up, above methods.

dhlaluku replied:
Done

nvazquez commented on engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:
```diff
@@ -1112,6 +1110,13 @@ public void orchestrateStart(final String vmUuid, final Map<VirtualMachineProfil
                     vmGuru.finalizeDeployment(cmds, vmProfile, dest, ctx);
 
+                    // Get VM extraConfig from DB and set to VM TO
+                    UserVmDetailVO detailVO = userVmDetailsDao.findDetail(vm.getId(), "extraConfig");
```
How about renaming the variable to `extraConfigDetail` or similar? Also, I think we can use an static variable the detail name instead of using the string "extraConfig" each time we want to access it

dhlaluku replied:
Removed all instances of "extraConfig" and replaced with ApiConstants.EXTRA_CONFIG.
Also renamed detailVO to extraConfigDetail

nvazquez commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java:
```diff
@@ -2186,6 +2185,16 @@ So if getMinSpeed() returns null we fall back to getSpeed().
+            String extraConfig = vmTO.getExtraConfig();
+            if (extraConfig != null) {
+                genericXML.setExtraConfig(extraConfig);
+                vm.addComp(genericXML);
```
Nice. Have you tested adding the component without the genericXML wrapper? I think I saw it supported strings

dhlaluku replied:
Correct, I tried it and it works just as the same.

nvazquez commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtVMDef.java:
```diff
@@ -1834,4 +1834,17 @@ public String toString() {
         return vmBuilder.toString();
     }
+
+    public static class GenericXML {
```
In case the previous comment works, this class won't be needed

dhlaluku replied:
Removing redundant class

nvazquez commented on server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:
```diff
@@ -4852,9 +4857,39 @@ public UserVm createVirtualMachine(DeployVMCmd cmd) throws InsufficientCapacityE
+
+                if (vm.getType() == VirtualMachine.Type.User) {
+                    validateConfigOption(decodedUrl, callerId);
+                    _uservmDetailsDao.addDetail(vm.getId(), "extraConfig", decodedUrl, true);
```
As mentioned before, the "extraConfig" could be replaced by a static variable

dhlaluku replied:
Done

nvazquez commented on server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:
```diff
@@ -4852,9 +4857,39 @@ public UserVm createVirtualMachine(DeployVMCmd cmd) throws InsufficientCapacityE
+            if (!Strings.isNullOrEmpty(extraConfig)) {
+                String decodedUrl;
+                try {
+                    decodedUrl = URLDecoder.decode(extraConfig, "UTF-8");
```
Nice. How about moving this logic into a new method, say 'decodeExtraConfig', which returns the decoded string or throws the exception

dhlaluku replied:
Refactored logic to a new method and removed hard-coded "extraConfig"

nvazquez commented on test/integration/smoke/test_vm_additional_config.py:
```diff
@@ -0,0 +1,130 @@
+        respone = self.apiclient.deployVirtualMachine(cmd)
+
+
+        #
```
Please remove commented code

dhlaluku replied:
Removed

nvazquez commented on server/src/test/java/com/cloud/vm/UserVmManagerTest.java:
```diff
@@ -1055,4 +1058,9 @@ public void testValidateUrlEncodedBase64() throws UnsupportedEncodingException {
     }
+
+    @Test
+    public void testCreateVirtualMachineWithAdditionalConfig() throws Exception {
```
If no work is done on this yet, it can be removed

nvazquez commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java:
```diff
@@ -2186,6 +2185,14 @@ So if getMinSpeed() returns null we fall back to getSpeed().
+        // Add extra configuration to User VM Domain XML before starting
+        if (vmTO.getType() == VirtualMachine.Type.User) {
+            String extraConfig = vmTO.getExtraConfig();
+            if (!Strings.isNullOrEmpty(extraConfig)){
```
Can you replace this for `StringUtils.isNotBlank()' method on Apache Commons package?

dhlaluku replied:
Done

dhlaluku replied:
Done

nvazquez commented on server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:
```diff
@@ -4852,9 +4860,58 @@ public UserVm createVirtualMachine(DeployVMCmd cmd) throws InsufficientCapacityE
+    protected void validateConfigOption(String decodedUrl, Long callerId) {
+        String[] accessConfigBlackList = AdditionalConfigBlackList.valueIn(callerId).split(",");
+        for (String configOption : accessConfigBlackList) {
+            if (decodedUrl.contains(configOption)) {
```
I think at this place you should also split the decoded url and check if the any key matches the blacklist element (not contains). For example these fake scenarios will throw an exception as it is right now:

blacklist = cpu,mem
extraconfig: "key1=allowcpu"

blacklist = cpu,mem
extraconfig: "memory=1111"

dhlaluku replied:
Implemented separate parsing mechanisms per hypervisor.
kvm -> Check XML tags against blacklist
Vmware/Xen -> Check keys against blacklist

nvazquez commented on server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:
```diff
@@ -4852,9 +4860,58 @@ public UserVm createVirtualMachine(DeployVMCmd cmd) throws InsufficientCapacityE
+        Account caller = CallContext.current().getCallingAccount();
+        String extraConfig = cmd.getExtraConfig();
+        Long vmId = vm.getId();
+        if (!Strings.isNullOrEmpty(extraConfig)) {
```
StringUtils.isNotBlank suggested as above. Also, please add the condition of the EnableAdditionalVmConfig global setting here and remove it from the addExtraConfig() method below

dhlaluku replied:
Adapted change and moved global config setting

yadvr commented on api/src/main/java/org/apache/cloudstack/api/command/user/vm/DeployVMCmd.java:
```diff
@@ -200,6 +199,9 @@
             " an optional parameter used to create additional data disks from datadisk templates; can't be specified with diskOfferingId parameter")
     private Map dataDiskTemplateToDiskOfferingList;
 
+    @Parameter(name = ApiConstants.EXTRA_CONFIG, type = CommandType.STRING, since = "4.12", description = "an optional URL encoded string that can be passed to the virtual machine upon successful deployment", length = 5120)
```
The deploy VM command is executed as a `GET` request which usually has a max. size of 2048 (2k).

yadvr replied:
Is this API arg allowed for all users or just the RoleType.Admin (in which case add the `authorized` field)

dhlaluku replied:
Usage is allowed for all users controlled by Global config set by root admin

yadvr commented on engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:
```diff
@@ -1112,6 +1113,13 @@ public void orchestrateStart(final String vmUuid, final Map<VirtualMachineProfil
 
+                    // Get VM extraConfig from DB and set to VM TO
+                    UserVmDetailVO extraConfigDetail = userVmDetailsDao.findDetail(vm.getId(), ApiConstants.EXTRA_CONFIG);
+
```
nitpicking here - Remove extra newline.

dhlaluku replied:
done

yadvr commented on server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:
```diff
@@ -4852,9 +4874,105 @@ public UserVm createVirtualMachine(DeployVMCmd cmd) throws InsufficientCapacityE
+                        }
+                    }
+                }
+
```
nitpicking - remove newline.

dhlaluku replied:
Done

yadvr commented on server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:
```diff
@@ -4852,9 +4874,105 @@ public UserVm createVirtualMachine(DeployVMCmd cmd) throws InsufficientCapacityE
+                        }
+                    }
+                }
+
```
nitpicking - remove newline.

dhlaluku replied:
Done

yadvr commented on test/integration/smoke/test_vm_additional_config.py:
```diff
@@ -0,0 +1,143 @@
+        SshClient(host_ip, port=22, user=self.hostConfig["username"])
+
+    @attr(tags = ["devcloud", "advanced", "advancedns", "smoke", "basic", "sg"], required_hardware="false")
+    def test_deploy_vm_config(self):
```
Show result of the test, fix it to make it work a specific hypervisor as I understand each hypervisor will have it's own allowed `extraconfig` string. You can also add a negative and postivite test per hypervisor.

dhlaluku replied:
Tests updated for each hypervisor

yadvr reviewed (changes requested):
Overall good work Dingane, left some comments and TODOs.

yadvr commented:
@dhlaluku can you also rebase against latest master and fix conflicts.

nvazquez commented on server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:
```diff
@@ -2393,18 +2409,24 @@ public UserVm updateVirtualMachine(UpdateVMCmd cmd) throws ResourceUnavailableEx
+                vmInstance.setDetails(details);
+                _vmDao.saveDetails(vmInstance);
+            }
+            if (!Strings.isNullOrEmpty(extraConfig)) {
```
Will need to check for 'EnableAdditionalVmConfig' value as well (account id can be retrieved from vmInstance). Also, can you replace this for StringUtils.isNotBlank() ?

dhlaluku replied:
Updated

nvazquez commented on server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:
```diff
@@ -2393,18 +2409,24 @@ public UserVm updateVirtualMachine(UpdateVMCmd cmd) throws ResourceUnavailableEx
+                _vmDao.saveDetails(vmInstance);
+            }
+            if (!Strings.isNullOrEmpty(extraConfig)) {
+                Account account = CallContext.current().getCallingAccount();
```
You can use accountDao at this point to get the account by ID (you can get it from vmInstance.getAccountId())

Commit:
Suqash commits to a single commit and rebase against master

Update marvin tests to use white list

borisstoyanov commented on test/integration/smoke/test_vm_extra_config_data.py:
```diff
@@ -0,0 +1,299 @@
+                'Extra  configuration not found in instance vmx file'
+            )
+
+        self.destroy_vm(response.id)
```
I think we're missing a test for Xen Server @dhlaluku

dhlaluku replied:
Let me add them quickly

borisstoyanov commented on test/integration/smoke/test_vm_extra_config_data.py:
```diff
@@ -0,0 +1,299 @@
+# Import System modules
+import urllib
+import xml.etree.ElementTree as ET
+from lxml import etree
```
I don't think this is part of stock marvin installation, it will fail when we try to run smoketests against it.

dhlaluku replied:
They should be part of the python standard library, I do not remember if I had to pip install them but I think they come as default.

borisstoyanov replied:
not in my setup, I have a python with just Marvin and it's dependencies and PyCharm ide cannot resolve lxml

dhlaluku replied:
Is the a workaround?

borisstoyanov replied:
I've just checked with a marvin installation and it's there, looks like it's missing on Mac, or we add it during the setup. Let leave this for now

borisstoyanov commented on test/integration/smoke/test_deploy_vm_extra_config_data.py:
```diff
@@ -294,3 +300,67 @@ def test_deploy_vm__with_extraconfig_vmware(self):
+                result = ssh_client.execute(vm_config_check)
+                param_value = config_tuple[1].strip()
+                if param_value == '""':
+                    result[0] = '""';
```
Maybe the semicolon here is not needed

dhlaluku replied:
True, I will remove them. I was probably mixing Java with Python.

borisstoyanov commented on test/integration/smoke/test_deploy_vm_extra_config_data.py:
```diff
@@ -0,0 +1,361 @@
+        self.destroy_vm(response.id)
+
+    @attr(tags=["devcloud", "advanced", "advancedns", "smoke", "basic", "sg"], required_hardware="true")
+    def test_update_vm__with_extraconfig_kvm(self):
```
there's 2 tests methods with the same name, if they are testing different things maybe it's good idea to change the names and add description as comment

borisstoyanov commented on test/integration/smoke/test_deploy_vm_extra_config_data.py:
```diff
@@ -0,0 +1,361 @@
+
+    @attr(tags=["devcloud", "advanced", "advancedns", "smoke", "basic", "sg"], required_hardware="true")
+    def test_deploy_vm__with_extraconfig_vmware(self):
+        if self.hypervisor.lower() != 'vmware':
```
Add test description, what it does and what are the assertions that it makes

dhlaluku replied:
Added test description, please review and advise if it needs more details

borisstoyanov commented on test/integration/smoke/test_deploy_vm_extra_config_data.py:
```diff
@@ -0,0 +1,361 @@
+
+    @attr(tags=["devcloud", "advanced", "advancedns", "smoke", "basic", "sg"], required_hardware="true")
+    def test_deploy_vm__with_extraconfig_xenserver(self):
+        if self.hypervisor.lower() != 'xenserver':
```
Add test description, what it does and what are the assertions that it makes

dhlaluku replied:
Added test description, please review and advise if it needs more details

borisstoyanov commented on test/integration/smoke/test_deploy_vm_extra_config_data.py:
```diff
@@ -0,0 +1,360 @@
+        cmd.id = id
+        return self.apiclient.destroyVirtualMachine(cmd)
+
+    def deploy_vm(self, hypervisor, extra_config=None):
```
VMs are left over after the execution, we'll need to append the newly created VM to cls.cleanup

dhlaluku replied:
Added code to append vm to cleanup

borisstoyanov commented on test/integration/smoke/test_deploy_vm_extra_config_data.py:
```diff
@@ -0,0 +1,360 @@
+            self.destroy_vm(response.id)
+
+    @attr(tags=["devcloud", "advanced", "advancedns", "smoke", "basic", "sg"], required_hardware="true")
+    def test_deploy_vm__with_extraconfig_vmware(self):
```
This test fails with the following error:
`test_deploy_vm__with_extraconfig_vmware (tests.smoke.test_deploy_vm_extra_config_data.TestAddConfigtoDeployVM): CRITICAL: EXCEPTION: test_deploy_vm__with_extraconfig_vmware: ['Traceback (most recent call last):\n', '  File "/usr/lib64/python2.7/unittest/case.py", line 369, in run\n    testMethod()\n', '  File "/marvin/tests/smoke/test_deploy_vm_extra_config_data.py", line 289, in test_deploy_vm__with_extraconfig_vmware\n    extraconfig_decoded = urllib.unquote(extraconfig)\n', '  File "/usr/lib64/python2.7/urllib.py", line 1217, in unquote\n    bits = s.split(\'%\')\n', "AttributeError: 'NoneType' object has no attribute 'split'\n"]
--------------------- >> end captured logging << ---------------------`

dhlaluku replied:
Fixed and updated test for vmware

dhlaluku replied:
===SSH to Host 10.2.2.207 port : 22 SUCCESSFUL===
{Cmd: cat "$(esxcli vm process list | grep i-2-55-VM | tail -1 | awk '{print $3}')" | grep -w 'hypervisor.cpuid.v0 = "FALSE"' via Host: 10.2.2.207} {returns: [u'hypervisor.cpuid.v0 = "FALSE"']}
=== TestName: test_deploy_vm__with_extraconfig_vmware | Status : SUCCESS ===

borisstoyanov commented on test/integration/smoke/test_deploy_vm_extra_config_data.py:
```diff
@@ -0,0 +1,360 @@
+        self.destroy_vm(response.id)
+
+    @attr(tags=["devcloud", "advanced", "advancedns", "smoke", "basic", "sg"], required_hardware="true")
+    def test_deploy_vm__with_extraconfig_xenserver(self):
```
this test fails with the following error: 
```
[root@trl-1748-k-M7-bstoyanov-marvin ~]# cat /marvin//MarvinLogs/test_deploy_vm_extra_config_data_KGOSEO/failed_plus_exceptions.txt
2018-12-06 09:11:05,135 - CRITICAL - FAILED: test_deploy_vm__with_extraconfig_xenserver: ['Traceback (most recent call last):\n', '  File "/usr/lib64/python2.7/unittest/case.py", line 369, in run\n    testMethod()\n', '  File "/marvin/tests/smoke/test_deploy_vm_extra_config_data.py", line 343, in test_deploy_vm__with_extraconfig_xenserver\n    \'Extra  configuration not found in VM param list\'\n', '  File "/usr/lib64/python2.7/unittest/case.py", line 553, in assertEqual\n    assertion_func(first, second, msg=msg)\n', '  File "/usr/lib64/python2.7/unittest/case.py", line 546, in _baseAssertEqual\n    raise self.failureException(msg)\n', 'AssertionError: Extra  configuration not found in VM param list\n']

```

dhlaluku replied:
Fixed and updated test

====Trying SSH Connection: Host:10.2.2.241 User:root                                   Port:22 RetryCnt:60===
===SSH to Host 10.2.2.241 port : 22 SUCCESSFUL===
{Cmd: xe vm-list name-label=i-2-52-VM params=uuid  via Host: 10.2.2.241} {returns: [u'uuid ( RO)    : 7a8c243e-c8d4-1419-2811-7c73fc8ffe7d', u'', u'']}
{Cmd: xe vm-param-get param-name=PV-bootloader uuid=7a8c243e-c8d4-1419-2811-7c73fc8ffe7d via Host: 10.2.2.241} {returns: [u'pygrub']}
{Cmd: xe vm-param-get param-name=PV-args uuid=7a8c243e-c8d4-1419-2811-7c73fc8ffe7d via Host: 10.2.2.241} {returns: [u'hvc0']}
{Cmd: xe vm-param-get param-name=HVM-boot-policy uuid=7a8c243e-c8d4-1419-2811-7c73fc8ffe7d via Host: 10.2.2.241} {returns: [u'']}
=== TestName: test_deploy_vm__with_extraconfig_xenserver | Status : SUCCESS ===

borisstoyanov reviewed (changes requested):
There's quite a few comments in the marvin test @dhlaluku, can you please look into them?

borisstoyanov commented on test/integration/smoke/test_deploy_vm_extra_config_data.py:
```diff
@@ -0,0 +1,360 @@
+        self.apiclient = self.testClient.getApiClient()
+        self.hypervisor = self.testClient.getHypervisorInfo()
+        self.dbclient = self.testClient.getDbConnection()
+        self.cleanup = []
```
Isn't cleanup already instantiated on a class level?

dhlaluku replied:
True, I have removed the cleanup here

borisstoyanov commented on test/integration/smoke/test_deploy_vm_extra_config_data.py:
```diff
@@ -0,0 +1,360 @@
+        config_arr = extraconfig_decoded.splitlines()
+
+        for config in config_arr:
+            grep_config = "cat /vmfs/volumes/eaa90eaf-b789d201/%s/%s.vmx | grep -w '%s'" % (
```
hard-coded?

borisstoyanov replied:
you can try using the vmware client on the host to get the info about the vmx file: 
[root@trl-1748-k-M7-bstoyanov-esxi1:~] esxcli vm process list
```
i-2-26-VM
   World ID: 77590
   Process ID: 0
   VMX Cartel ID: 77589
   UUID: 42 06 ce 56 19 ff dd 22-7d a9 00 8c 79 85 e0 09
   Display Name: i-2-26-VM
   Config File: /vmfs/volumes/dc7ba692-c6334326/i-2-26-VM/i-2-26-VM.vmx
```

dhlaluku replied:
fixed

borisstoyanov commented:
https://cwiki.apache.org/confluence/display/CLOUDSTACK/Enable+additional+configuration+metadata+to+virtual+machines

dhlaluku commented:
@nvazquez @DaanHoogland, please review

dhlaluku commented:
@nvazquez @DaanHoogland @rhtyd may you please have a look at this PR. @rhtyd you requested changes a while back when you reviewed this PR, please have another and advice on any revisions required.

Commit:
* Fix marvin test failure
* Add new marvin negative tests cases
* Remove hard-coded hypervisor types in marvin tests

borisstoyanov commented on test/integration/smoke/test_deploy_vm_extra_config_data.py:
```diff
@@ -0,0 +1,522 @@
+                host = list_hosts(
+                    self.apiclient,
+                    id=host_id,
+                    hypervisor='kvm')
```
This shouldn't be 'kvm' hardcoded, but instead use 'testClient.getHypervisorInfo()' to set a class var. In this case hypervisors !=kvm will run

dhlaluku replied:
Done, now using `self.hypervisor.lower()`

nvazquez committed:
ui: Fix SystemVMs public range dedication (#3495)

Fix small UI issue when dedicating public IR range for system VMs:
Unable to execute API command createvlaniprange due to invalid value. Invalid parameter domainid value=undefined due to incorrect long value format, or entity does not exist or due to incorrect parameter annotation for the field in api cmd class.

Fixes: #3485

nvazquez committed:
KVM: Enhancements for direct download feature (#3374)

* Add revoke certificates API

* Add background task to sync certificates

* Fix marvin test and revoke certificate

* Fix certificate sent to hypervisor was missing headers

* Fix background task for uploading certificates to hosts

yadvr committed:
Allow users to share templates with Accounts or Projects through the UI

* Allow users to share templates with Accounts or Projects through the
updateTemplate permissions API

* Change behaviour to show only supported projects and accounts with update template permissions

* Allow admins to see accounts dropdown and only hide lists for users

* Don't allow sharing project owned templates as you cannot retrieve them in list api calls

Commit:
server: disable unauthenticated integration.api.port by default (#3504)

Set integration.api.port to (0) zero as default. CloudStack provides CloudStack API Unauthenticated Access through port 8096. It should not be open to the Internet in any case.

yadvr committed:
vmware: fix volume stats logic (#3473)

During volume stats calculation, if a volume has more than one disk in
the chain-info it is not used to sum the physical and virtual size
in the loop, instead any previous entry was overwritten by the last disk.

Signed-off-by: Rohit Yadav <rohit.yadav@shapeblue.com>

nvazquez committed:
server: Fix hardcoded max data volumes when VM has been created but not started before (#3494)

When VM is created but not started on any host, attaching a volume on it causes the maximum number of allowed volumes to be 6 (hardcoded).

Commit:
server: fix the subnet overlap checking logic for tagged and untagged vlans when adding ipranges (#3430)

Fixes: #3114

When adding iprange for VLANs there are 3 cases -

VLAN under consideration has a tag (like 101)
VLAN under consideration has a tag but as a range (like 101-124)
VLAN is untagged (i.e. id is "untagged")
Before adding iprange we have to check for possible overlaps and throw exception. This needs to be done as follows -

If VLAN Tag ID is numeric or a range we need to call UriUtils.checkVlanUriOverlapmethod which internally tries to expand the range as verifies if there are overlaps. If URI overlaps (i.e. there are overlapping VLAN tags) we then need to verify if the iprange being added overlaps with previously added ranges.
If there are no overlapping tags we simply need to test for public networks being present in the VLAN.
A Regression was introduced in 41fdb88#diff-6e2b61984e8fa2823bb47da3caafa4eeR3174 which caused comparing 'untagged' string as a numeric VLAN Tag range and and attempted expanding it to test overlap in UriUtils.checkVlanUriOverlap.

To fix the bug in the issue, we need to handle the untagged case separately as it's non-numeric tag in code. For untagged VLANs and overlapping VLAN URIs we need to check for ipranges and gateways which happens naturally after this change. For tagged VLANs with non-overlapping URIs we need to check if there is a public network.

borisstoyanov commented:
@rhtyd @nvazquez @shwstppr @anuragaw, this is FR04 from the roadmap items, it needs to be rebased and there's merge conflicts.

Commit:
Fix build error after rebase and add hugepagesless

anuragaw commented:
Closing this and opening a public PR as advised.
